File: src/schema.rs
//! Database schema.

diesel::table! {
    media (id) {
        id -> Binary,
        kind -> Text,
        created -> BigInt,
        updated -> BigInt,
    }
}

diesel::table! {
    tags (media, label) {
        media -> Binary,
        label -> Text,
    }
}

diesel::table! {
    logs (id) {
        id -> Binary,
        media -> Binary,
        kind -> Text,
        date -> BigInt,
    }
}

diesel::table! {
    books (id) {
        id -> Binary,
        isbn -> Nullable<Text>,
        hcid -> Nullable<BigInt>,
        title -> Text,
        cover -> Nullable<Text>,
        about -> Nullable<Text>,
        color -> Nullable<Text>,
    }
}

diesel::table! {
    films (id) {
        id -> Binary,
        tmdb -> Nullable<BigInt>,
        title -> Text,
        year -> Nullable<BigInt>,
        rating -> Nullable<BigInt>,
    }
}

diesel::table! {
    games (id) {
        id -> Binary,
        title -> Text,
        system -> Nullable<Text>,
        rating -> Nullable<BigInt>,
    }
}

diesel::table! {
    links (id) {
        id -> Binary,
        url -> Text,
        title -> Nullable<Text>,
    }
}

diesel::table! {
    shows (id) {
        id -> Binary,
        tmdb -> Nullable<BigInt>,
        title -> Text,
        year -> Nullable<BigInt>,
        rating -> Nullable<BigInt>,
    }
}

diesel::table! {
    games_systems (id) {
        id -> Binary,
        title -> Text,
        system -> Nullable<Text>,
        region -> Nullable<Text>,
        model -> Nullable<Text>,
        revision -> Nullable<Text>,
        serial -> Nullable<Text>,
        variant -> Nullable<Text>,
        complete -> Bool,
        modified -> Bool,
    }
}

diesel::table! {
    games_copies (id) {
        id -> Binary,
        title -> Nullable<Text>,
        system -> Nullable<Text>,
        region -> Nullable<Text>,
        model -> Nullable<Text>,
        revision -> Nullable<Text>,
        serial -> Nullable<Text>,
        complete -> Bool,
        modified -> Bool,
    }
}

diesel::table! {
    games_copies_ref (copy, game) {
        copy -> Binary,
        game -> Binary,
        idx -> BigInt,
    }
}

diesel::table! {
    games_systems_ref (system, game) {
        system -> Binary,
        game -> Binary,
        idx -> BigInt,
    }
}

diesel::table! {
    games_extras (id) {
        id -> Binary,
        title -> Text,
        system -> Nullable<Text>,
        region -> Nullable<Text>,
        model -> Nullable<Text>,
        revision -> Nullable<Text>,
        serial -> Nullable<Text>,
        variant -> Nullable<Text>,
        complete -> Bool,
        modified -> Bool,
    }
}

diesel::table! {
    games_extras_ref (extra, game) {
        extra -> Binary,
        game -> Binary,
        idx -> BigInt,
    }
}

diesel::joinable!(books -> media (id));
diesel::joinable!(films -> media (id));
diesel::joinable!(games -> media (id));
diesel::joinable!(games_copies_ref -> games (game));
diesel::joinable!(games_extras_ref -> games_extras (extra));
diesel::joinable!(games_systems_ref -> games_systems (system));
diesel::joinable!(games_copies_ref -> games_copies (copy));
diesel::joinable!(links -> media (id));
diesel::joinable!(logs -> media (media));
diesel::joinable!(shows -> media (id));
diesel::joinable!(tags -> media (media));

diesel::allow_tables_to_appear_in_same_query!(
    books,
    films,
    games,
    games_copies,
    games_copies_ref,
    games_extras,
    games_extras_ref,
    games_systems,
    games_systems_ref,
    links,
    logs,
    shows,
    media,
    tags,
);

File: src/sql/schema.ts
//! Database schema.
//!
//! Each table is exported under its own name, in the order it appears in the
//! schema.

import {
    customType,
    integer,
    primaryKey,
    sqliteTable,
    text,
} from "drizzle-orm/sqlite-core";

/** UUIDs are stored as blobs and exposed as hyphenated lowercase text. */
const uuid = customType<{ data: string; driverData: Buffer }>({
    dataType: () => "blob",
    toDriver: (id) => Buffer.from(id.replaceAll("-", ""), "hex"),
    fromDriver: (raw) => {
        const hex = Buffer.from(raw).toString("hex");
        return [
            hex.slice(0, 8),
            hex.slice(8, 12),
            hex.slice(12, 16),
            hex.slice(16, 20),
            hex.slice(20),
        ].join("-");
    },
});

/** Booleans are stored as the integers zero and one. */
const flag = () => integer({ mode: "boolean" }).notNull();

export const media = sqliteTable("media", {
    id: uuid().primaryKey(),
    kind: text().notNull(),
    created: integer().notNull(),
    updated: integer().notNull(),
});

export const tags = sqliteTable(
    "tags",
    {
        media: uuid().notNull(),
        label: text().notNull(),
    },
    (self) => [primaryKey({ columns: [self.media, self.label] })],
);

export const logs = sqliteTable("logs", {
    id: uuid().primaryKey(),
    media: uuid().notNull(),
    kind: text().notNull(),
    date: integer().notNull(),
});

export const books = sqliteTable("books", {
    id: uuid().primaryKey(),
    isbn: text(),
    hcid: integer(),
    title: text().notNull(),
    cover: text(),
    about: text(),
    color: text(),
});

export const books_owned = sqliteTable("books_owned", {
    id: uuid().primaryKey(),
    isbn: text().notNull(),
    title: text(),
    edition: text(),
});

export const books_author = sqliteTable(
    "books_author",
    {
        book: uuid().notNull(),
        name: text().notNull(),
        idx: integer().notNull(),
    },
    (self) => [primaryKey({ columns: [self.book, self.name] })],
);

export const links = sqliteTable("links", {
    id: uuid().primaryKey(),
    url: text().notNull(),
    title: text(),
});

export const games = sqliteTable("games", {
    id: uuid().primaryKey(),
    title: text().notNull(),
    platform: text(),
    rating: integer(),
});

export const games_owned = sqliteTable("games_owned", {
    id: uuid().primaryKey(),
    kind: text().notNull(),
    title: text(),
    platform: text(),
    region: text(),
    model: text(),
    revision: text(),
    serial: text(),
    variant: text(),
    complete: flag(),
    modified: flag(),
});

export const games_owned_ref = sqliteTable(
    "games_owned_ref",
    {
        owned: uuid().notNull(),
        game: uuid().notNull(),
        idx: integer().notNull(),
    },
    (self) => [primaryKey({ columns: [self.owned, self.game] })],
);

export const films = sqliteTable("films", {
    id: uuid().primaryKey(),
    tmdb: integer(),
    title: text().notNull(),
    year: integer(),
    rating: integer(),
});

export const shows = sqliteTable("shows", {
    id: uuid().primaryKey(),
    tmdb: integer(),
    title: text().notNull(),
    year: integer(),
    rating: integer(),
});
